teamcity: failed test: TestBatchTimeout

The following tests appear to have failed on master (testrace): TestBatchTimeout, TestBatchTimeout/NoTimeout, TestBatchTimeout/WithTimeout

You may want to check [for open issues](https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue+is%3Aopen+TestBatchTimeout).

[#1389505](https://teamcity.cockroachdb.com/viewLog.html?buildId=1389505):


```
TestBatchTimeout/NoTimeout
--- FAIL: testrace/TestBatchTimeout/NoTimeout (0.000s)
Test ended in panic.

------- Stdout: -------
I190716 05:13:10.789280 133 util/stop/stopper.go:542  quiescing; tasks left:
2      [async] 
panic: test timed out after 45m0s

goroutine 185 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/go/src/testing/testing.go:1334 +0x11c
created by time.goFunc
	/usr/local/go/src/time/sleep.go:169 +0x52

goroutine 1 [chan receive, 44 minutes]:
testing.(*T).Run(0xc0001b2300, 0x15a64a5, 0x10, 0x16273d0, 0x1)
	/usr/local/go/src/testing/testing.go:917 +0x693
testing.runTests.func1(0xc0001b2300)
	/usr/local/go/src/testing/testing.go:1157 +0xa9
testing.tRunner(0xc0001b2300, 0xc000503db0)
	/usr/local/go/src/testing/testing.go:865 +0x164
testing.runTests(0xc00026ee40, 0x237b120, 0xb, 0xb, 0x0)
	/usr/local/go/src/testing/testing.go:1155 +0x524
testing.(*M).Run(0xc000474200, 0x0)
	/usr/local/go/src/testing/testing.go:1072 +0x2ec
main.main()
	_testmain.go:62 +0x223

goroutine 20 [syscall, 45 minutes]:
os/signal.signal_recv(0x478031)
	/usr/local/go/src/runtime/sigqueue.go:139 +0x9c
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:23 +0x30
created by os/signal.init.0
	/usr/local/go/src/os/signal/signal_unix.go:29 +0x4f

goroutine 21 [chan receive]:
github.com/cockroachdb/cockroach/pkg/util/log.flushDaemon()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:1222 +0x14f
created by github.com/cockroachdb/cockroach/pkg/util/log.init.0
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:625 +0x1e4

goroutine 22 [chan receive, 45 minutes]:
github.com/cockroachdb/cockroach/pkg/util/log.signalFlusher()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:632 +0x14b
created by github.com/cockroachdb/cockroach/pkg/util/log.init.0
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:626 +0x1fc

goroutine 133 [semacquire, 44 minutes]:
sync.runtime_Semacquire(0xc0000e8848)
	/usr/local/go/src/runtime/sema.go:56 +0x39
sync.(*WaitGroup).Wait(0xc0000e8840)
	/usr/local/go/src/sync/waitgroup.go:130 +0xb2
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc0000e8820, 0x184a880, 0xc0000bc010)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:489 +0x43d
github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.TestBatchTimeout(0xc000652100)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher_test.go:375 +0x250
testing.tRunner(0xc000652100, 0x16273d0)
	/usr/local/go/src/testing/testing.go:865 +0x164
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:916 +0x65b

goroutine 184 [select, 44 minutes]:
github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.chanSender.Send(0xc000540420, 0x184a900, 0xc0001ec0f0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher_test.go:48 +0x1a5
github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.(*RequestBatcher).sendBatch.func1(0x184a900, 0xc0001ec0f0, 0xc0002b20e0, 0xc0001acbd0)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:265 +0x126
github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher.(*RequestBatcher).sendBatch.func3(0x184a900, 0xc0001ec0f0)
	/go/src/github.com/cockroachdb/cockroach/pkg/internal/client/requestbatcher/batcher.go:279 +0xc4
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc00027a020, 0xc0000e8820, 0xc0001ec090)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:196 +0x160
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:189 +0xc4



TestBatchTimeout/WithTimeout
--- FAIL: testrace/TestBatchTimeout/WithTimeout (0.000s)
Test ended in panic.




TestBatchTimeout
--- FAIL: testrace/TestBatchTimeout (0.000s)
Test ended in panic.





```

Please assign, take a look and update the issue accordingly.

--- a/pkg/internal/client/requestbatcher/batcher.go
+++ b/pkg/internal/client/requestbatcher/batcher.go
@@ -336,6 +336,11 @@ func (b *RequestBatcher) cleanup(err error) {
 }
 
 func (b *RequestBatcher) run(ctx context.Context) {
+	// Create a context to be used in sendBatch to cancel in-flight batches when
+	// this function exits. If we did not cancel in-flight requests then the
+	// Stopper might get stuck waiting for those requests to complete.
+	sendCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
 	var (
 		// inFlight tracks the number of batches currently being sent.
 		// true.
@@ -359,7 +364,7 @@ func (b *RequestBatcher) run(ctx context.Context) {
 			if inFlight >= b.cfg.InFlightBackpressureLimit {
 				inBackPressure = true
 			}
-			b.sendBatch(ctx, ba)
+			b.sendBatch(sendCtx, ba)
 		}
 		handleSendDone = func() {
 			inFlight--

--- a/pkg/internal/client/requestbatcher/batcher_test.go
+++ b/pkg/internal/client/requestbatcher/batcher_test.go
@@ -44,14 +44,18 @@ type chanSender chan batchSend
 func (c chanSender) Send(
 	ctx context.Context, ba roachpb.BatchRequest,
 ) (*roachpb.BatchResponse, *roachpb.Error) {
-	respChan := make(chan batchResp)
+	respChan := make(chan batchResp, 1)
 	select {
 	case c <- batchSend{ctx: ctx, ba: ba, respChan: respChan}:
 	case <-ctx.Done():
 		return nil, roachpb.NewError(ctx.Err())
 	}
-	resp := <-respChan
-	return resp.br, resp.pe
+	select {
+	case resp := <-respChan:
+		return resp.br, resp.pe
+	case <-ctx.Done():
+		return nil, roachpb.NewError(ctx.Err())
+	}
 }
 
 func TestBatcherSendOnSizeWithReset(t *testing.T) {
@@ -261,25 +265,29 @@ func TestSendAfterCanceled(t *testing.T) {
 	assert.Equal(t, err, ctx.Err())
 }
 
+// TestStopDuringSend ensures that in-flight requests are canceled when the
+// RequestBatcher's stopper indicates that it should quiesce.
 func TestStopDuringSend(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	stopper := stop.NewStopper()
 	sc := make(chanSender, 1)
 	b := New(Config{
-		Sender:  sc,
-		Stopper: stopper,
-		MaxWait: 10 * time.Millisecond,
-		MaxIdle: 10 * time.Millisecond,
+		Sender:          sc,
+		Stopper:         stopper,
+		MaxMsgsPerBatch: 1,
 	})
 	errChan := make(chan error)
 	go func() {
 		_, err := b.Send(context.Background(), 1, &roachpb.GetRequest{})
 		errChan <- err
 	}()
-	r := <-sc
-	go stopper.Stop(context.Background())
-	assert.Equal(t, <-errChan, stop.ErrUnavailable)
-	r.respChan <- batchResp{}
+	// Wait for the request to get sent.
+	<-sc
+	stopper.Stop(context.Background())
+	// Depending on the ordering of when channels close the sender might
+	// get one of two errors.
+	assert.True(t, testutils.IsError(<-errChan,
+		stop.ErrUnavailable.Error()+"|"+context.Canceled.Error()))
 }
 
 func TestPanicWithNilStopper(t *testing.T) {
@@ -354,15 +362,21 @@ func TestBatchTimeout(t *testing.T) {
 		var wg sync.WaitGroup
 		wg.Add(2)
 		var err1, err2 error
-		go func() { _, err1 = b.Send(ctx1, 1, &roachpb.GetRequest{}); wg.Done() }()
+		err1Chan := make(chan error, 1)
+		go func() {
+			_, err1 = b.Send(ctx1, 1, &roachpb.GetRequest{})
+			err1Chan <- err1
+			wg.Done()
+		}()
 		go func() { _, err2 = b.Send(ctx2, 1, &roachpb.GetRequest{}); wg.Done() }()
 		select {
 		case s := <-sc:
 			assert.Len(t, s.ba.Requests, 2)
 			s.respChan <- batchResp{}
-		case <-ctx1.Done():
+		case <-err1Chan:
 			// This case implies that the test did not exercise what was intended
 			// but that's okay, clean up the other request and return.
+			assert.Equal(t, context.DeadlineExceeded, err1)
 			t.Logf("canceled goroutine failed to send within %v, passing", timeout)
 			cancel2()
 			wg.Wait()

